roachtest: add RequiresDeprecatedWorkload testSpec

DarrylWong opened the pull request:
Previously we required the workload binary to be present for all tests to run even if the test did not use it.

Now that we have removed all unnecessary usage of the deprecated workload, we can switch the validation of the workload binary to be opt in.

Fixes:  #128172, #127954
Epic: none
Release note: none

DarrylWong committed:
roachtest: add RequiresDeprecatedWorkload testSpec

Previously we required the workload binary to be
present for all tests to run even if the test
did not use it.

Now that we have removed all uncessary usage
of the deprecated workload, we can switch the
validation of the workload binary to be opt in.

Co-authored-by: dt <david@cockroachlabs.com>

Fixes: none
Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127837)
<!-- Reviewable:end -->

DarrylWong commented:
Copied general approach from here: https://github.com/cockroachdb/cockroach/pull/119521/commits/b26c93bbc9db60bb7cd2e25fb8ab2e1fb497d745, but added in a hard fail to stop usage + updated the set of tests that need the testSpec.

nameisbhaskar commented on pkg/cmd/roachtest/test_impl.go:
```diff
@@ -195,6 +195,11 @@ func (t *testImpl) StandardCockroach() string {
 func (t *testImpl) DeprecatedWorkload() string {
+	// Fail and give a hint that the test should set `RequiresDeprecatedWorkload` so
+	// we can validate the binary exists.
+	if !t.spec.RequiresDeprecatedWorkload {
```
Can we just fail here if `deprecatedWorkload` is "" instead of introducing "RequiresDeprecatedWorkload"?

DarrylWong replied:
My comment is badly worded, I'll update it. I'm not guarding against the binary not existing, I'm guarding against people using the deprecated workload when not needed.

In a previous PR, I refactored almost all the `DeprecatedWorkload()` calls to just use `cockroach workload` instead. I want future tests to stop using `DeprecatedWorkload` if possible, so if a test does need it I want the test to explicitly set `RequiresDeprecatedWorkload`. 

If we just check that a `deprecatedWorkload` exists, it would be possible for a test writer to not set this spec and continue using the `deprecatedWorkload` without even knowing it's discouraged.

Let me know if that makes sense.

csgourav commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -730,6 +730,11 @@ func (r *testRunner) runWorker(
 		}
 
+		// Verify that the deprecated workload is available if needed.
+		if testToRun.spec.RequiresDeprecatedWorkload && workload[arch] == "" {
```
Can we do this earlier? Maybe after test selection, if there is any test with RequiresDeprecatedWorkload, fail early or remove those from selected tests list with warning, as opposed to checking this for all the tests.

DarrylWong replied:
I just copied the approach of how we verify that other binaries, i.e. `libgeos`, exist when the test marks them as required.

I agree that it'd be nice if it could know earlier and skip/fail them right at the start though. Let me look into it! 😃

DarrylWong replied:
Ah, the reason it can't be earlier is because we need to know the arch of the test, which isn't picked until right above here: 

https://github.com/cockroachdb/cockroach/blob/f91b32003f394af57d7acbda3a6f0776045c4c40/pkg/cmd/roachtest/test_runner.go#L704-L721

renatolabs commented:
Would it make sense to auto-upload the `workload` binary to the workload node if the spec uses the new field?

DarrylWong commented:
>Would it make sense to auto-upload the workload binary to the workload node if the spec uses the new field?

Good idea, I'll add it! It doesn't cover *all* cases unfortunately so we have to keep `DeprecatedWorkload` public.

DarrylWong committed:
roachtest: auto upload workload binary if required by workload node

Currently, if a test wishes to use the deprecated workload, they must
both set the RequiresDeprecatedWorkload test spec and manually upload
it in the test itself.

However, if the test provisions a workload node via the WorkloadNode
cluster spec, we know that the test most likely wants the workload there.
This change checks that the above two conditions are true, and now
automatically uploads the workload binary for the test.

This leaves just 3 occurences of tests manually uploading the workload
binary themselves.

renatolabs reviewed (approved):
Nice!

DarrylWong commented:
[Passing run](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/16296274?buildTab=tests) of tests that use the deprecated binary

TFTR!

bors r+

DarrylWong commented:
bors cancel

DarrylWong commented:
Snuck in a quick fix for the tpchbench tests that use `querybench`, which need the deprecated binary.

DarrylWong commented:
bors r+

DarrylWong commented:
blathers backport 24.2.0-rc
